From 05c3f7f8b934f956db063f56d5dcf12c4e69d314 Mon Sep 17 00:00:00 2001
From: Huynh Thanh Hung <hung.huynh.wz@rvc.renesas.com>
Date: Thu, 11 Oct 2018 15:29:57 +0700
Subject: [PATCH] gstplaybin change vspfilter as default converter

Signed-off-by: Huynh Thanh Hung <hung.huynh.wz@rvc.renesas.com>
---
 configure.ac | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/configure.ac b/configure.ac
index b5d8366..27a82a2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -609,6 +609,10 @@ AG_GST_CHECK_FEATURE(XSHM, [X Shared Memory extension], [ ], [
   AC_SUBST(XSHM_LIBS)
 ])

+dnl set default video converter
+DEFAULT_CNV="vspfilter"
+AC_SUBST(DEFAULT_CNV)
+
 dnl *** ext plug-ins ***
 dnl keep this list sorted alphabetically !

--
2.7.4
