From 1365b15777baa02e44654d27b3661934bd634552 Mon Sep 17 00:00:00 2001
From: Huynh Thanh Hung <hung.huynh.wz@rvc.renesas.com>
Date: Thu, 11 Oct 2018 15:32:48 +0700
Subject: [PATCH] convertframe: change videoconvert to vspfilter

Signed-off-by: Huynh Thanh Hung <hung.huynh.wz@rvc.renesas.com>
---
 gst-libs/gst/video/convertframe.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/gst-libs/gst/video/convertframe.c b/gst-libs/gst/video/convertframe.c
index 61e9c1a..9c5b697 100644
--- a/gst-libs/gst/video/convertframe.c
+++ b/gst-libs/gst/video/convertframe.c
@@ -123,7 +123,7 @@ build_convert_frame_pipeline (GstElement ** src_element,
       g_warning
           ("build_convert_frame_pipeline: Buffer has crop metadata but videocrop element is not found. Cropping will be disabled");
     } else {
-      if (!create_element ("videoconvert", &csp2, &error))
+      if (!create_element ("vspfilter", &csp2, &error))
         goto no_elements;
     }
   }
@@ -131,12 +131,12 @@ build_convert_frame_pipeline (GstElement ** src_element,
   /* videoscale is here to correct for the pixel-aspect-ratio for us */
   GST_DEBUG ("creating elements");
   if (!create_element ("appsrc", &src, &error) ||
-      !create_element ("videoconvert", &csp, &error) ||
+      !create_element ("vspfilter", &csp, &error) ||
       !create_element ("videoscale", &vscale, &error) ||
       !create_element ("appsink", &sink, &error))
     goto no_elements;

-  pipeline = gst_pipeline_new ("videoconvert-pipeline");
+  pipeline = gst_pipeline_new ("vspfilter-pipeline");
   if (pipeline == NULL)
     goto no_pipeline;

--
2.7.4
