From b15d52c6cc38a3834f6505f0d6bd31095140acc3 Mon Sep 17 00:00:00 2001
From: Quoc Ngo <quoc.ngo.wz@rvc.renesas.com>
Date: Wed, 29 Aug 2018 19:36:28 +0700
Subject: [PATCH] demo3d: add h264parse to correct pipeline play video on
 demo3d

Signed-off-by: Quoc Ngo <quoc.ngo.wz@rvc.renesas.com>
---
 Demo3D.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Demo3D.c b/Demo3D.c
index 65c3d76..99853b7 100644
--- a/Demo3D.c
+++ b/Demo3D.c
@@ -792,7 +792,7 @@ touch_handle(void *data, wl_fixed_t x_w, wl_fixed_t y_w)
           if (gv_init) {            
             /*initialie Gstreamer video pipeline*/
             gv_init = 0;
-            sprintf(gst_video.string, "filesrc location=/home/root/Demo3D/Renesas.mp4 ! qtdemux ! omxh264dec ! appsink caps=video/x-raw,format=NV12,width=%d,height=%d name=appsink", video_w, video_h);  
+            sprintf(gst_video.string, "filesrc location=/home/root/Demo3D/Renesas.mp4 ! qtdemux ! h264parse ! omxh264dec ! appsink caps=video/x-raw,format=NV12,width=%d,height=%d name=appsink", video_w, video_h);  
             gst_video.gstapp_new_sample = gstapp_new_sample;
             gst_video.gst_bus_message = gst_bus_message;
             gst_video.type = 0; //video 
-- 
1.9.1

